### DISCUSSION TITLE
Introduce Useful Features for Interface Inference

### DISCUSSION THREAD

**Discussion by DUXINYI** [dotnet/csharplang#5930] [Language Ideas]
During creating domain data model I found it hard to get a way that is simple, readable and error-sensible enough to describe the internal relation among so many Interfaces.
I need a stronger Inference System to deal with static data constraint.

I think there are several useful features that could be introduced into Interface Inference System:

- Literal Types
For example in TypeScript: 
`type Female = 'female';`
There is a similar mechanism which is called Enum in C#, however, it only supports integer value mapping.

- Property-based Interface Equivalent Determination
For example in TypeScript, interface i1 is compatible with interface i2:
```
interface i1{
  prop1:string;
}
interface i2{
  prop1:string;
}
```
Considering that C# Interface might contains Method Memer, this situation only fits on "Pure Property" Interfaces.

- Aggregatable Interfaces (Property Omittable)
For example in TypeScript:
```
interface OnlyWidth: Omit<Rectable,'height'>{
}
```
which is equal to:
```
interface OnlyWidth{
  width:number;
}
```
Sometimes define an entirely entity is better than a splitted entity, and in many time we use "omit" to express "I don't need" but not "it doesn't exist".

- (Primary) Property Value Constraint Support

Like data constraint such as:

- - Numeric Operator Constraint
`public int len where len > 0; //len>0 satisfied all the time`
- - Regex Constraint
`public string birthday where match @"^\d{4}-\d{2}-\d{2}$";`
- - Convert between Number Literal and String Literal
`public string lenStr where int(lenStr) < 3;`
- - Aggregated Constraint
`public int len where len > 0 && len < 100;`

**Comment by TahirAhmadov** [comment]
These things were requested previously. String-backed enums were requested, which would be a runtime and language change. I don't recall which discussion/issue it is. (Anybody?)
For the "property omitting" / "imitation inheritance" where you "copy" some members of a type to another type, but not all, without actual inheritance, please see https://github.com/dotnet/csharplang/discussions/5862.
For the property value constraint, see https://github.com/dotnet/csharplang/issues/3630.

**Reply by DUXINYI** [reply]
Should I follow discussion of these topics instead of create a new one? Alright. Thanks your reply

**Reply by TahirAhmadov** [reply]
You should review the existing items and choose either to voice your support, suggest changes, or clarify how what you need is completely different (in case I misunderstood).

**Reply by DUXINYI** [reply]
#5862 proposed applying "property omitting" / "imitation inheritance" to class type instead of interface. In my case, I hope it work on interface and retain the integrity of class.

#3630 only checks null parameters however I need more detailed value control as the content body I updated.

**Comment by huoyaoyuan** [comment]
The type system of JavaScript is way different from .NET. Interfaces are just assertions of dictionaries in TypeScript.

> Property-based Interface Equivalent Determination
For example in TypeScript, interface i1 is compatible with interface i2:

In C# they are not, and they can have different implementations on same type.

**Reply by DUXINYI** [reply]
I don't want to break the current equivalent determination mechanism, for existed modules they still work as what they did.

What I want is in some case "strongly matched" determination doesn't matter, programmar has chance to switch into a simple way to design the interfaces. Just like what anonymous type do.

**Reply by huoyaoyuan** [reply]
It should suite better with shapes. Interfaces have strong runtime meaning, and codegen requirements.
